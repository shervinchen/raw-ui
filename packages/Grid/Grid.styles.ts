import { isPlainObject, isString } from 'lodash';
import {
  AlignResponsiveStyle,
  Aligns,
  Align,
  BreakPoints,
  ColResponsiveStyle,
  ColSpanStyle,
  ColStyle,
  GutterResponsiveStyle,
  Justifies,
  JustifyResponsiveStyle,
  Justify,
  ResponsiveValue,
} from './Grid.types';

const GRID_SIZE = 24;

export const alignStyleMap: {
  [key in Align]: string;
} = {
  normal: 'normal',
  top: 'flex-start',
  center: 'center',
  bottom: 'flex-end',
};

export const justifyStyleMap: {
  [key in Justify]: string;
} = {
  normal: 'normal',
  start: 'flex-start',
  center: 'center',
  end: 'flex-end',
  'space-between': 'space-between',
  'space-around': 'space-around',
  'space-evenly': 'space-evenly',
};

const isLegalGutterNumber = (value: ResponsiveValue<number>) => {
  return Number.isInteger(value) && (value as number) >= 0;
};

const isLegalColPropertyValue = (value: ResponsiveValue<number>) => {
  return (
    Number.isInteger(value) &&
    (value as number) >= 1 &&
    (value as number) <= GRID_SIZE
  );
};

const isLegalGridPropertyValue = (
  type: 'gutter' | 'align' | 'justify',
  value:
    | ResponsiveValue<number>
    | ResponsiveValue<Align>
    | ResponsiveValue<Justify>
) => {
  const keys = {
    align: Aligns,
    justify: Justifies,
  };
  return type === 'gutter'
    ? isLegalGutterNumber(value as ResponsiveValue<number>)
    : isString(value) && Object.keys(keys[type]).includes(value);
};

const calculateColStyle = (
  type: 'span' | 'offset' | 'order',
  value: number
) => {
  return {
    span: {
      display: 'block',
      flex: `0 0 ${(value / GRID_SIZE) * 100}%`,
      maxWidth: `${(value / GRID_SIZE) * 100}%`,
    },
    offset: {
      marginInlineStart: `${(value / GRID_SIZE) * 100}%`,
    },
    order: {
      order: value,
    },
  }[type];
};

export const getColResponsiveStyle = (
  type: 'span' | 'offset' | 'order',
  value: ResponsiveValue<number>
) => {
  const defaultStyle: ColStyle = {
    span: {
      display: 'block',
      flex: '0 1 auto',
      maxWidth: 'none',
    },
    offset: {
      marginInlineStart: 0,
    },
    order: {
      order: 0,
    },
  };
  const hideSpanStyle: ColSpanStyle = {
    display: 'none',
    flex: '0 1 auto',
    maxWidth: 'none',
  };

  const responsiveStyle: ColResponsiveStyle = {
    xs: defaultStyle[type],
    sm: defaultStyle[type],
    md: defaultStyle[type],
    lg: defaultStyle[type],
    xl: defaultStyle[type],
    xxl: defaultStyle[type],
  };

  if (type === 'span' && value === 0) {
    Object.keys(responsiveStyle).map((key) => {
      responsiveStyle[key] = hideSpanStyle;
    });
    return responsiveStyle;
  }

  if (isLegalColPropertyValue(value)) {
    Object.keys(responsiveStyle).map((key) => {
      responsiveStyle[key] = calculateColStyle(type, value as number);
    });
    return responsiveStyle;
  }

  if (isPlainObject(value)) {
    const breakPoints = Object.keys(BreakPoints);
    Object.keys(value).map((breakPoint) => {
      if (type === 'span' && value[breakPoint] === 0) {
        responsiveStyle[breakPoint] = hideSpanStyle;
      } else if (
        breakPoints.includes(breakPoint) &&
        isLegalColPropertyValue(value[breakPoint])
      ) {
        responsiveStyle[breakPoint] = calculateColStyle(
          type,
          value[breakPoint]
        );
      }
    });
    return responsiveStyle;
  }

  return responsiveStyle;
};

export const getGridResponsiveStyle = (
  type: 'gutter' | 'align' | 'justify',
  value:
    | ResponsiveValue<number>
    | ResponsiveValue<Align>
    | ResponsiveValue<Justify>
) => {
  const gridResponsiveStyle:
    | GutterResponsiveStyle
    | AlignResponsiveStyle
    | JustifyResponsiveStyle =
    type === 'gutter'
      ? {
          xs: 0,
          sm: 0,
          md: 0,
          lg: 0,
          xl: 0,
          xxl: 0,
        }
      : {
          xs: 'normal',
          sm: 'normal',
          md: 'normal',
          lg: 'normal',
          xl: 'normal',
          xxl: 'normal',
        };

  if (isLegalGridPropertyValue(type, value)) {
    Object.keys(gridResponsiveStyle).map((key) => {
      gridResponsiveStyle[key] = value;
    });
    return gridResponsiveStyle;
  }

  if (isPlainObject(value)) {
    const breakPoints = Object.keys(BreakPoints);
    Object.keys(value).map((breakPoint) => {
      if (
        breakPoints.includes(breakPoint) &&
        isLegalGridPropertyValue(type, value[breakPoint])
      ) {
        gridResponsiveStyle[breakPoint] = value[breakPoint] as
          | number
          | Align
          | Justify;
      }
    });
    return gridResponsiveStyle;
  }

  return gridResponsiveStyle;
};
