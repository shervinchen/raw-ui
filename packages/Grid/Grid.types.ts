import { CSSProperties, HTMLAttributes } from 'react';

export enum BreakPoints {
  xs = 'xs',
  sm = 'sm',
  md = 'md',
  lg = 'lg',
  xl = 'xl',
  xxl = 'xxl',
}
export type BreakPoint = keyof typeof BreakPoints;

export type ResponsiveValue<T> = T | { [key in BreakPoint]?: T };

export type Gutter = [ResponsiveValue<number>, ResponsiveValue<number>];

export enum Aligns {
  normal = 'normal',
  top = 'top',
  center = 'center',
  bottom = 'bottom',
}
export enum Justifies {
  normal = 'normal',
  start = 'start',
  center = 'center',
  end = 'end',
  'space-between' = 'space-between',
  'space-around' = 'space-around',
  'space-evenly' = 'space-evenly',
}

export type AlignValue = keyof typeof Aligns;
export type JustifyValue = keyof typeof Justifies;

interface BaseGridProps {
  gutter?: Gutter;
  align?: ResponsiveValue<AlignValue>;
  justify?: ResponsiveValue<JustifyValue>;
  className?: string;
}

interface BaseGridColProps {
  offset?: ResponsiveValue<number>;
  order?: ResponsiveValue<number>;
  span?: ResponsiveValue<number>;
  className?: string;
  style?: CSSProperties;
}

type NativeGridProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  keyof BaseGridProps
>;

type NativeGridColProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  keyof BaseGridColProps
>;

export type GridProps = BaseGridProps & NativeGridProps;
export type GridColProps = BaseGridColProps & NativeGridColProps;

export interface ColSpanStyle {
  display: string;
  flex: string;
  maxWidth: string;
}

export interface ColOffsetStyle {
  marginInlineStart: number | string;
}

export interface ColOrderStyle {
  order: number;
}

export interface ColStyle {
  span: ColSpanStyle;
  offset: ColOffsetStyle;
  order: ColOrderStyle;
}

export type ColSpanResponsiveStyle = {
  [key in BreakPoint]: ColSpanStyle;
};

export type ColOffsetResponsiveStyle = {
  [key in BreakPoint]: ColOffsetStyle;
};

export type ColOrderResponsiveStyle = {
  [key in BreakPoint]: ColOrderStyle;
};

export type ColResponsiveStyle = {
  [key in BreakPoint]: ColSpanStyle | ColOffsetStyle | ColOrderStyle;
};

export type GutterResponsiveStyle = {
  [key in BreakPoint]: number;
};

export interface GridConfig {
  horizontalGutterResponsiveStyle: GutterResponsiveStyle;
}

export type AlignResponsiveStyle = {
  [key in BreakPoint]: AlignValue;
};

export type JustifyResponsiveStyle = {
  [key in BreakPoint]: JustifyValue;
};
