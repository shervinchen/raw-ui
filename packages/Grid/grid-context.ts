import { createContext, useContext } from 'react';
import { GridConfig } from './Grid.types';

const defaultContext: GridConfig = {
  horizontalGutter: 0,
};

export const GridContext = createContext<GridConfig>(defaultContext);

export const useGridContext = (): GridConfig =>
  useContext<GridConfig>(GridContext);
