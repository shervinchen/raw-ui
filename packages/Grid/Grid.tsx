import React, { FC, PropsWithChildren } from 'react';
import classNames from 'classnames';
import {
  AlignResponsiveStyle,
  GridConfig,
  GridProps,
  GutterResponsiveStyle,
  JustifyResponsiveStyle,
} from './Grid.types';
import { GridContext } from './grid-context';
import {
  alignStyleMap,
  getGridResponsiveStyle,
  justifyStyleMap,
} from './Grid.styles';

const Grid: FC<PropsWithChildren<GridProps>> = ({
  gutter = [0, 0],
  align = 'normal',
  justify = 'normal',
  className = '',
  children,
  ...restProps
}) => {
  const classes = classNames(
    'raw-grid',
    'raw-grid-gutter-side',
    'raw-grid-gutter-vertical',
    'raw-grid-align',
    'raw-grid-justify',
    className
  );

  const horizontalGutterResponsiveStyle = getGridResponsiveStyle(
    'gutter',
    gutter?.[0]
  ) as GutterResponsiveStyle;
  const verticalGutterResponsiveStyle = getGridResponsiveStyle(
    'gutter',
    gutter?.[1]
  ) as GutterResponsiveStyle;
  const alignResponsiveStyle = getGridResponsiveStyle(
    'align',
    align
  ) as AlignResponsiveStyle;
  const justifyResponsiveStyle = getGridResponsiveStyle(
    'justify',
    justify
  ) as JustifyResponsiveStyle;

  const gridConfig: GridConfig = {
    horizontalGutterResponsiveStyle,
  };

  return (
    <GridContext.Provider value={gridConfig}>
      <div className={classes} {...restProps}>
        {children}
        <style jsx>{`
          .raw-grid {
            display: flex;
            flex-flow: row wrap;
          }
          .raw-grid-gutter-side {
            margin-left: -${horizontalGutterResponsiveStyle.xs / 2}px;
            margin-right: -${horizontalGutterResponsiveStyle.xs / 2}px;
          }
          .raw-grid-gutter-vertical {
            row-gap: ${verticalGutterResponsiveStyle.xs}px;
          }
          .raw-grid-align {
            align-items: ${alignStyleMap[alignResponsiveStyle.xs]};
          }
          .raw-grid-justify {
            justify-content: ${justifyStyleMap[justifyResponsiveStyle.xs]};
          }
          @media (min-width: 576px) {
            .raw-grid-gutter-side {
              margin-left: -${horizontalGutterResponsiveStyle.sm / 2}px;
              margin-right: -${horizontalGutterResponsiveStyle.sm / 2}px;
            }
            .raw-grid-gutter-vertical {
              row-gap: ${verticalGutterResponsiveStyle.sm}px;
            }
            .raw-grid-align {
              align-items: ${alignStyleMap[alignResponsiveStyle.sm]};
            }
            .raw-grid-justify {
              justify-content: ${justifyStyleMap[justifyResponsiveStyle.sm]};
            }
          }
          @media (min-width: 768px) {
            .raw-grid-gutter-side {
              margin-left: -${horizontalGutterResponsiveStyle.md / 2}px;
              margin-right: -${horizontalGutterResponsiveStyle.md / 2}px;
            }
            .raw-grid-gutter-vertical {
              row-gap: ${verticalGutterResponsiveStyle.md}px;
            }
            .raw-grid-align {
              align-items: ${alignStyleMap[alignResponsiveStyle.md]};
            }
            .raw-grid-justify {
              justify-content: ${justifyStyleMap[justifyResponsiveStyle.md]};
            }
          }
          @media (min-width: 992px) {
            .raw-grid-gutter-side {
              margin-left: -${horizontalGutterResponsiveStyle.lg / 2}px;
              margin-right: -${horizontalGutterResponsiveStyle.lg / 2}px;
            }
            .raw-grid-gutter-vertical {
              row-gap: ${verticalGutterResponsiveStyle.lg}px;
            }
            .raw-grid-align {
              align-items: ${alignStyleMap[alignResponsiveStyle.lg]};
            }
            .raw-grid-justify {
              justify-content: ${justifyStyleMap[justifyResponsiveStyle.lg]};
            }
          }
          @media (min-width: 1200px) {
            .raw-grid-gutter-side {
              margin-left: -${horizontalGutterResponsiveStyle.xl / 2}px;
              margin-right: -${horizontalGutterResponsiveStyle.xl / 2}px;
            }
            .raw-grid-gutter-vertical {
              row-gap: ${verticalGutterResponsiveStyle.xl}px;
            }
            .raw-grid-align {
              align-items: ${alignStyleMap[alignResponsiveStyle.xl]};
            }
            .raw-grid-justify {
              justify-content: ${justifyStyleMap[justifyResponsiveStyle.xl]};
            }
          }
          @media (min-width: 1600px) {
            .raw-grid-gutter-side {
              margin-left: -${horizontalGutterResponsiveStyle.xxl / 2}px;
              margin-right: -${horizontalGutterResponsiveStyle.xxl / 2}px;
            }
            .raw-grid-gutter-vertical {
              row-gap: ${verticalGutterResponsiveStyle.xxl}px;
            }
            .raw-grid-align {
              align-items: ${alignStyleMap[alignResponsiveStyle.xxl]};
            }
            .raw-grid-justify {
              justify-content: ${justifyStyleMap[justifyResponsiveStyle.xxl]};
            }
          }
        `}</style>
      </div>
    </GridContext.Provider>
  );
};

export default Grid;
