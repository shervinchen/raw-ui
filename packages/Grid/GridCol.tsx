import React, { FC, PropsWithChildren } from 'react';
import classNames from 'classnames';
import {
  ColOffsetResponsiveStyle,
  ColOrderResponsiveStyle,
  ColSpanResponsiveStyle,
  GridColProps,
} from './Grid.types';
import { useGridContext } from './grid-context';
import { getColResponsiveStyle } from './Grid.styles';

const GridCol: FC<PropsWithChildren<GridColProps>> = ({
  offset = 0,
  order = 0,
  span,
  className = '',
  children,
  ...restProps
}) => {
  const { horizontalGutterResponsiveStyle } = useGridContext();

  const classes = classNames(
    'raw-grid-col',
    'raw-grid-col-span',
    'raw-grid-col-offset',
    'raw-grid-col-order',
    'raw-grid-gutter-horizontal',
    className
  );

  const colSpanResponsiveStyle = getColResponsiveStyle(
    'span',
    span
  ) as ColSpanResponsiveStyle;
  const colOffsetResponsiveStyle = getColResponsiveStyle(
    'offset',
    offset
  ) as ColOffsetResponsiveStyle;
  const colOrderResponsiveStyle = getColResponsiveStyle(
    'order',
    order
  ) as ColOrderResponsiveStyle;

  return (
    <div data-testid="gridCol" className={classes} {...restProps}>
      {children}
      <style jsx>{`
        .raw-grid-col {
          position: relative;
          max-width: 100%;
          min-height: 1px;
        }
        .raw-grid-col-span {
          display: ${colSpanResponsiveStyle.xs.display};
          flex: ${colSpanResponsiveStyle.xs.flex};
          max-width: ${colSpanResponsiveStyle.xs.maxWidth};
        }
        .raw-grid-col-offset {
          margin-inline-start: ${colOffsetResponsiveStyle.xs.marginInlineStart};
        }
        .raw-grid-col-order {
          order: ${colOrderResponsiveStyle.xs.order};
        }
        .raw-grid-gutter-horizontal {
          padding-left: ${horizontalGutterResponsiveStyle.xs / 2}px;
          padding-right: ${horizontalGutterResponsiveStyle.xs / 2}px;
        }
        @media (min-width: 576px) {
          .raw-grid-col-span {
            display: ${colSpanResponsiveStyle.sm.display};
            flex: ${colSpanResponsiveStyle.sm.flex};
            max-width: ${colSpanResponsiveStyle.sm.maxWidth};
          }
          .raw-grid-col-offset {
            margin-inline-start: ${colOffsetResponsiveStyle.sm
              .marginInlineStart};
          }
          .raw-grid-col-order {
            order: ${colOrderResponsiveStyle.sm.order};
          }
          .raw-grid-gutter-horizontal {
            padding-left: ${horizontalGutterResponsiveStyle.sm / 2}px;
            padding-right: ${horizontalGutterResponsiveStyle.sm / 2}px;
          }
        }
        @media (min-width: 768px) {
          .raw-grid-col-span {
            display: ${colSpanResponsiveStyle.md.display};
            flex: ${colSpanResponsiveStyle.md.flex};
            max-width: ${colSpanResponsiveStyle.md.maxWidth};
          }
          .raw-grid-col-offset {
            margin-inline-start: ${colOffsetResponsiveStyle.md
              .marginInlineStart};
          }
          .raw-grid-col-order {
            order: ${colOrderResponsiveStyle.md.order};
          }
          .raw-grid-gutter-horizontal {
            padding-left: ${horizontalGutterResponsiveStyle.md / 2}px;
            padding-right: ${horizontalGutterResponsiveStyle.md / 2}px;
          }
        }
        @media (min-width: 992px) {
          .raw-grid-col-span {
            display: ${colSpanResponsiveStyle.lg.display};
            flex: ${colSpanResponsiveStyle.lg.flex};
            max-width: ${colSpanResponsiveStyle.lg.maxWidth};
          }
          .raw-grid-col-offset {
            margin-inline-start: ${colOffsetResponsiveStyle.lg
              .marginInlineStart};
          }
          .raw-grid-col-order {
            order: ${colOrderResponsiveStyle.lg.order};
          }
          .raw-grid-gutter-horizontal {
            padding-left: ${horizontalGutterResponsiveStyle.lg / 2}px;
            padding-right: ${horizontalGutterResponsiveStyle.lg / 2}px;
          }
        }
        @media (min-width: 1200px) {
          .raw-grid-col-span {
            display: ${colSpanResponsiveStyle.xl.display};
            flex: ${colSpanResponsiveStyle.xl.flex};
            max-width: ${colSpanResponsiveStyle.xl.maxWidth};
          }
          .raw-grid-col-offset {
            margin-inline-start: ${colOffsetResponsiveStyle.xl
              .marginInlineStart};
          }
          .raw-grid-col-order {
            order: ${colOrderResponsiveStyle.xl.order};
          }
          .raw-grid-gutter-horizontal {
            padding-left: ${horizontalGutterResponsiveStyle.xl / 2}px;
            padding-right: ${horizontalGutterResponsiveStyle.xl / 2}px;
          }
        }
        @media (min-width: 1600px) {
          .raw-grid-col-span {
            display: ${colSpanResponsiveStyle.xxl.display};
            flex: ${colSpanResponsiveStyle.xxl.flex};
            max-width: ${colSpanResponsiveStyle.xxl.maxWidth};
          }
          .raw-grid-col-offset {
            margin-inline-start: ${colOffsetResponsiveStyle.xxl
              .marginInlineStart};
          }
          .raw-grid-col-order {
            order: ${colOrderResponsiveStyle.xxl.order};
          }
          .raw-grid-gutter-horizontal {
            padding-left: ${horizontalGutterResponsiveStyle.xxl / 2}px;
            padding-right: ${horizontalGutterResponsiveStyle.xxl / 2}px;
          }
        }
      `}</style>
    </div>
  );
};

export default GridCol;
