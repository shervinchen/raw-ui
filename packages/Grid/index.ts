import Grid from './Grid';
import GridCol from './GridCol';

export type GridComponentType = typeof Grid & {
  Col: typeof GridCol;
};

(Grid as GridComponentType).Col = GridCol;

export default Grid as GridComponentType;
