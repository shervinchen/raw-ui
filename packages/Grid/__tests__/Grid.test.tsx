import React from 'react';
import { render, screen } from '@testing-library/react';
import Grid from '..';

describe('Grid', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(
      <Grid>
        <Grid.Col>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should support custom class name', () => {
    render(
      <Grid className="custom-grid">
        <Grid.Col className="custom-grid-col">
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getByTestId('grid')).toHaveClass('custom-grid');
    expect(screen.getByTestId('gridCol')).toHaveClass('custom-grid-col');
  });

  test('should support set regular value of gutter', () => {
    render(
      <Grid gutter={[16, 24]}>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getByTestId('grid')).toHaveStyle({
      rowGap: '24px',
      marginLeft: '-8px',
      marginRight: '-8px',
    });
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      paddingLeft: '8px',
      paddingRight: '8px',
    });
  });

  test('should support set responsive value of gutter', () => {
    render(
      <Grid
        gutter={[
          { xs: 8, sm: 16, md: 24, lg: 32, xl: 40, xxl: 48 },
          { xs: 8, sm: 16, md: 24, lg: 32, xl: 40, xxl: 48 },
        ]}
      >
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getByTestId('grid')).toHaveStyle({
      rowGap: '8px',
      marginLeft: '-4px',
      marginRight: '-4px',
    });
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      paddingLeft: '4px',
      paddingRight: '4px',
    });
  });

  test('should support set regular align items', () => {
    render(
      <Grid align="top">
        <Grid.Col>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getByTestId('grid')).toHaveStyle({
      alignItems: 'flex-start',
    });
  });

  test('should support set responsive align items', () => {
    render(
      <Grid
        align={{
          xs: 'top',
          sm: 'center',
          md: 'bottom',
          lg: 'normal',
          xl: 'normal',
          xxl: 'normal',
        }}
      >
        <Grid.Col>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getByTestId('grid')).toHaveStyle({
      alignItems: 'flex-start',
    });
  });

  test('should support set regular justify content', () => {
    render(
      <Grid justify="start">
        <Grid.Col>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getByTestId('grid')).toHaveStyle({
      justifyContent: 'flex-start',
    });
  });

  test('should support set responsive justify content', () => {
    render(
      <Grid
        justify={{
          xs: 'start',
          sm: 'center',
          md: 'end',
          lg: 'space-around',
          xl: 'space-between',
          xxl: 'space-evenly',
        }}
      >
        <Grid.Col>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getByTestId('grid')).toHaveStyle({
      justifyContent: 'flex-start',
    });
  });

  test('should support set regular span of grid col', () => {
    render(
      <Grid>
        <Grid.Col span={24}>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      flex: '0 0 100%',
      maxWidth: '100%',
    });
  });

  test('should support set responsive span of grid col', () => {
    render(
      <Grid>
        <Grid.Col span={{ xs: 2, sm: 8, md: 4, lg: 10, xl: 6, xxl: 6 }}>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      flex: `0 0 ${(2 / 24) * 100}%`,
      maxWidth: `${(2 / 24) * 100}%`,
    });
  });

  test('should support set 0 span of grid col', () => {
    const { rerender } = render(
      <Grid>
        <Grid.Col span={0}>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      display: 'none',
    });
    rerender(
      <Grid>
        <Grid.Col span={{ xs: 0, sm: 0, md: 0, lg: 0, xl: 0, xxl: 0 }}>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      display: 'none',
    });
  });

  test('should support set regular offset of grid col', () => {
    render(
      <Grid>
        <Grid.Col span={8} offset={8}>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      marginInlineStart: `${(8 / 24) * 100}%`,
    });
  });

  test('should support set responsive offset of grid col', () => {
    render(
      <Grid>
        <Grid.Col offset={{ xs: 2, sm: 8, md: 4, lg: 10, xl: 6, xxl: 6 }}>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      marginInlineStart: `${(2 / 24) * 100}%`,
    });
  });

  test('should support set regular order of grid col', () => {
    render(
      <Grid>
        <Grid.Col span={6} order={4}>
          <div>col-order-4</div>
        </Grid.Col>
        <Grid.Col span={6} order={3}>
          <div>col-order-3</div>
        </Grid.Col>
        <Grid.Col span={6} order={2}>
          <div>col-order-2</div>
        </Grid.Col>
        <Grid.Col span={6} order={1}>
          <div>col-order-1</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      order: 4,
    });
    expect(screen.getAllByTestId('gridCol')[1]).toHaveStyle({
      order: 3,
    });
    expect(screen.getAllByTestId('gridCol')[2]).toHaveStyle({
      order: 2,
    });
    expect(screen.getAllByTestId('gridCol')[3]).toHaveStyle({
      order: 1,
    });
  });

  test('should support set responsive order of grid col', () => {
    render(
      <Grid>
        <Grid.Col span={6} order={{ xs: 4, sm: 1, md: 2, lg: 3 }}>
          <div>col-order-4</div>
        </Grid.Col>
        <Grid.Col span={6} order={{ xs: 3, sm: 2, md: 4, lg: 1 }}>
          <div>col-order-3</div>
        </Grid.Col>
        <Grid.Col span={6} order={{ xs: 2, sm: 3, md: 1, lg: 4 }}>
          <div>col-order-2</div>
        </Grid.Col>
        <Grid.Col span={6} order={{ xs: 1, sm: 4, md: 3, lg: 2 }}>
          <div>col-order-1</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      order: 4,
    });
    expect(screen.getAllByTestId('gridCol')[1]).toHaveStyle({
      order: 3,
    });
    expect(screen.getAllByTestId('gridCol')[2]).toHaveStyle({
      order: 2,
    });
    expect(screen.getAllByTestId('gridCol')[3]).toHaveStyle({
      order: 1,
    });
  });

  test('should get default style of grid when value is illegal', () => {
    const { rerender } = render(
      <Grid
        gutter={[-1, -1]}
        align={'unknown' as unknown}
        justify={'unknown' as unknown}
      >
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getByTestId('grid')).toHaveStyle({
      rowGap: '0px',
      marginLeft: '-0px',
      marginRight: '-0px',
      alignItems: 'normal',
      justifyContent: 'normal',
    });
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      paddingLeft: '0px',
      paddingRight: '0px',
    });
    rerender(
      <Grid
        gutter={[{ xs: -1 }, { xs: -1 }]}
        align={{ xs: 'unknown' } as unknown}
        justify={{ xs: 'unknown' } as unknown}
      >
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
        <Grid.Col span={6}>
          <div>col-6</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getByTestId('grid')).toHaveStyle({
      rowGap: '0px',
      marginLeft: '-0px',
      marginRight: '-0px',
      alignItems: 'normal',
      justifyContent: 'normal',
    });
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      paddingLeft: '0px',
      paddingRight: '0px',
    });
  });

  test('should get default style of grid col when value is illegal', () => {
    const { rerender } = render(
      <Grid>
        <Grid.Col span={-1} offset={-1} order={-1}>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      flex: '0 1 auto',
      maxWidth: 'none',
      marginInlineStart: 0,
      order: 0,
    });
    rerender(
      <Grid>
        <Grid.Col span={{ xs: -1 }} offset={{ xs: -1 }} order={{ xs: -1 }}>
          <div>col</div>
        </Grid.Col>
      </Grid>
    );
    expect(screen.getAllByTestId('gridCol')[0]).toHaveStyle({
      flex: '0 1 auto',
      maxWidth: 'none',
      marginInlineStart: 0,
      order: 0,
    });
  });
});
