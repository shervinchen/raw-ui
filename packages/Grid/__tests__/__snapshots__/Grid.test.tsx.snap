// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-3953341048 raw-grid raw-grid-gutter-side raw-grid-gutter-vertical raw-grid-align raw-grid-justify"
    data-testid="grid"
  >
    <div
      class="jsx-125401513 raw-grid-col raw-grid-col-span raw-grid-col-offset raw-grid-col-order raw-grid-gutter-horizontal"
      data-testid="gridCol"
    >
      <div>
        col
      </div>
    </div>
  </div>
</DocumentFragment>
`;
